<img src=".\images\Logo.png" width=150 align="left" /> <img src=".\images\Logo2.jpg" width=450 align="right" />


# <center><font color= #1e8449 > <b>CUxx_xxxxx</font></center>

<font color='steelblue'><b>Citizenlab Data Science Methodology > II - Data Processing Domain </font>
***
> # <font color='steelblue'> <b>05.- Data Collection</font>

<font color='steelblue'>Data Collection is the process to obtain and generate (if required) necessary data to model the problem.</font>

### 19. Simular datos diarios de infraestructuras

* Datos diarios para 2019
* Datos de ocupación y demanda de cada infraestructura cada día
* Datos de dos tipos de eventos: evento infraestructura y evento zona
* Hace falta identificar a las infraestructuras







<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Settings" data-toc-modified-id="Settings-1"><font color="green">Settings</font></a></span></li><li><span><a href="#Data-Load" data-toc-modified-id="Data-Load-2"><font color="green">Data Load</font></a></span></li><li><span><a href="#ETL-Processes" data-toc-modified-id="ETL-Processes-3">ETL Processes</a></span><ul class="toc-item"><li><span><a href="#Import-data-from:-CSV,-Excel,-Tab,-JSON,-SQL,-and-Parquet-files" data-toc-modified-id="Import-data-from:-CSV,-Excel,-Tab,-JSON,-SQL,-and-Parquet-files-3.1">Import data from: CSV, Excel, Tab, JSON, SQL, and Parquet files</a></span></li></ul></li><li><span><a href="#Synthetic-Data-Generation" data-toc-modified-id="Synthetic-Data-Generation-4">Synthetic Data Generation</a></span></li><li><span><a href="#Fake-Data-Generation" data-toc-modified-id="Fake-Data-Generation-5">Fake Data Generation</a></span></li><li><span><a href="#Open-Data" data-toc-modified-id="Open-Data-6">Open Data</a></span></li><li><span><a href="#Data-Save" data-toc-modified-id="Data-Save-7"><font color="green">Data Save</font></a></span></li><li><span><a href="#Main-Conclusions" data-toc-modified-id="Main-Conclusions-8"><font color=" #2874a6 ">Main Conclusions</font></a></span></li><li><span><a href="#Main-Actions" data-toc-modified-id="Main-Actions-9"><font color=" #2874a6 ">Main Actions</font></a></span><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#Acciones-done" data-toc-modified-id="Acciones-done-9.0.1"><font color=" #2874a6 ">Acciones done</font></a></span></li><li><span><a href="#Acctions-to-perform" data-toc-modified-id="Acctions-to-perform-9.0.2"><font color=" #2874a6 ">Acctions to perform</font></a></span></li></ul></li></ul></li></ul></div>

## <font color='green'>Settings</font>

### Packages to use


- {tcltk} para selección interactiva de archivos locales
- {tibble} para operaciones con data.frames
- {readr} para leer y escribir archivos csv
- {dplyr} para explorar datos


In [1]:
library(tibble)
library(readr)
library(dplyr)




Attaching package: 'dplyr'


The following objects are masked from 'package:stats':

    filter, lag


The following objects are masked from 'package:base':

    intersect, setdiff, setequal, union




### Paths

In [2]:
iPath <- "Data/Input/"
oPath <- "Data/Output/"

## <font color='green'>Data Load</font>

If there are more than one input file, make as many sections as files to import.

Instrucciones
- Los ficheros de entrada del proceso están siempre en Data/Input/.   
- Si hay más de un fichero de entrada, se crean tantos objetos iFile_xx y file_data_xx como ficheros de entrada (xx número correlativo con dos dígitos, rellenar con ceros a la izquierda)

<font color='tomato'><b> OPCION A:</b> Seleccionar fichero en ventana para mayor comodidad</font> 

Data load using the {tcltk} package. Ucomment the line if not using this option

In [3]:
# file_data <- tcltk::tk_choose.files(multi = FALSE)

<font color='tomato'><b> OPCION B:</b> Especificar el nombre de archivo</font>



In [4]:
iFile <- "CU_18_05_18_infraestructuras.csv"
file_data <- paste0(iPath, iFile)

if(file.exists(file_data)){
    cat("Se leerán datos del archivo: ", file_data)
} else{
    warning("Cuidado: el archivo no existe.")
}


Se leer<U+00E1>n datos del archivo:  Data/Input/CU_18_05_18_infraestructuras.csv

#### Data file to dataframe

Usar la función adecuada según el formato de entrada (xlsx, csv, json, ...)

In [5]:
data_01 <- read_csv(file_data)


Rows: 1138 Columns: 18
-- Column specification --------------------------------------------------------
Delimiter: ","
chr  (8): grupo, tipo, nombre, CODMUN, DIRECCION, info, CMUN, CDIS
dbl (10): dist_aeropuerto, dist_intercambiador, dist_cercanias, dist_hospita...

i Use `spec()` to retrieve the full column specification for this data.
i Specify the column types or set `show_col_types = FALSE` to quiet this message.


Estructura de  los datos:

In [6]:
glimpse(data_01)

Rows: 1,138
Columns: 18
$ grupo               <chr> "Transporte", "Transporte", "Transporte", "Transpo~
$ tipo                <chr> "Intercambiadores", "Intercambiadores", "Intercamb~
$ nombre              <chr> "Grandes Intercambiadores Plaza El<U+00ED>ptica", "Grande~
$ CODMUN              <chr> "079", "079", "079", "079", "079", "079", "079", "~
$ DIRECCION           <chr> "Plaza El<U+00ED>ptica s/n", "Calle Princesa, 89", "Estac~
$ info                <chr> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA~
$ CMUN                <chr> "079", "079", "079", "079", "079", "079", "079", "~
$ CDIS                <chr> "12", "09", "09", "05", "05", "05", "06", "02", "0~
$ dist_aeropuerto     <dbl> 5.2031430, 8.5507475, 7.3002463, 9.9339069, 9.1988~
$ dist_intercambiador <dbl> 0.0000000, 0.0000000, 0.0000000, 0.0000000, 0.0000~
$ dist_cercanias      <dbl> 1.648973148, 1.468638183, 0.088338692, 0.736877403~
$ dist_hospital       <dbl> 1.8154156, 0.4221115, 1.3855767, 1.0733484, 0.3488~
$ 

Muestra de datos:

In [7]:
slice_head(data_01, n = 5)

grupo      tipo            
1 Transporte Intercambiadores
2 Transporte Intercambiadores
3 Transporte Intercambiadores
4 Transporte Intercambiadores
5 Transporte Intercambiadores
  nombre                                              CODMUN
1 Grandes Intercambiadores Plaza El<U+00ED>ptica      079   
2 Grandes Intercambiadores Moncloa                    079   
3 Grandes Intercambiadores Pr<U+00ED>ncipe P<U+00ED>o 079   
4 Grandes Intercambiadores Plaza de Castilla          079   
5 Grandes Intercambiadores Avenida de Am<U+00E9>rica  079   
  DIRECCION                                  info CMUN CDIS dist_aeropuerto
1 Plaza El<U+00ED>ptica s/n                  NA   079  12   5.203143       
2 Calle Princesa, 89                         NA   079  09   8.550748       
3 Estaci<U+00F3>n Pr<U+00ED>ncipe P<U+00ED>o NA   079  09   7.300246       
4 Calle San Aquilino y Avenida de Asturias   NA   079  05   9.933907       
5 Avenida de Am<U+00E9>rica 9 A              NA   079  05   9.198897       
  dist_intercambiador dist_cercanias dist_hospital dist_universidad dist_gob 
1 0                   1.64897315     1.8154156     1.29960755       0.6177235
2 0                   1.46863818     0.4221115     0.30460781       0.6263847
3 0                   0.08833869     1.3855767     0.92076145       1.1700512
4 0                   0.73687740     1.0733484     0.08835165       0.1490496
5 0                   1.62310148     0.3488060     0.50695045       1.1979177
  dist_embajada dist_ccomercial X         Y       
1 0.39213238    0.7311571       -3.716577 40.38540
2 1.45739694    1.6105674       -3.719508 40.43474
3 5.62455975    0.9060857       -3.719560 40.42080
4 0.08316773    0.7758344       -3.689044 40.46719
5 0.40303107    0.9512580       -3.676731 40.43797

## ETL Processes

### Import data from: CSV, Excel, Tab, JSON, SQL, and Parquet files

Se han importado en el apartado Data Load anterior:

* Datos de infraestructuras

Incluir apartados si procede para: Extracción de datos (select, filter), Transformación de datos, (mutate, joins, ...). Si es necesario tratar datos perdidos, indicarlo también en NB 09.2

Si no aplica: Estos datos no requieren tareas de este tipo.

#### Data transform

* Crear identificador único para cada infraestructura

In [8]:
tdata_01 <- data_01 |>
    rowid_to_column("id_inf")

## Synthetic Data Generation

No aplica

## Fake Data Generation

Se generan datos de un año para cada una de las infraestructuras con las variables:

- capacidad
- demanda
- evento_infra
- evento_zona



* Generar fechas

In [9]:
inicio <- as.Date("2019-01-01")
fin <- as.Date("2019-12-31")
dias <- seq(inicio, fin, by = 1)


* Generar las filas suficientes

In [10]:
data_02 <- expand.grid(id_inf = tdata_01$id_inf,
            fecha = dias)

* Generar variables aleatoriamente

In [11]:
tdata_02 <- data_02 |>
    mutate(capacidad = rpois(n(), 1000)) |>
    mutate(demanda = capacidad + sample(c(-1, 1), n(), TRUE)*round(capacidad*0.1 + rnorm(n(), 10, 2))) |>
    rowwise() |>
    mutate(evento_infra = sample(c(0, 1), n(), TRUE, c((demanda > capacidad)*0.2 + (demanda < capacidad)*0.8, 
                                                        (demanda > capacidad)*0.8 + (demanda < capacidad)*0.2))) |>
    mutate(evento_zona = sample(c(0, 1), n(), TRUE, c((demanda > capacidad)*0.4 + (demanda < capacidad)*0.6, 
                                                        (demanda > capacidad)*0.6 + (demanda < capacidad)*0.4)))

                                                        

In [12]:
tdata_02 |> glimpse()

Rows: 415,370
Columns: 6
Rowwise: 
$ id_inf       <int> 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17~
$ fecha        <date> 2019-01-01, 2019-01-01, 2019-01-01, 2019-01-01, 2019-01-~
$ capacidad    <int> 935, 1067, 1030, 1010, 1016, 971, 994, 1023, 1010, 1001, ~
$ demanda      <dbl> 1039, 952, 1147, 900, 1125, 865, 886, 1136, 1121, 889, 11~
$ evento_infra <dbl> 1, 0, 1, 0, 0, 0, 0, 1, 1, 0, 1, 0, 1, 0, 0, 1, 0, 0, 0, ~
$ evento_zona  <dbl> 1, 1, 1, 0, 1, 1, 0, 1, 0, 0, 0, 1, 1, 0, 1, 0, 0, 1, 1, ~


In [13]:
tdata_02 |> slice_head(n = 5)

id_inf fecha      capacidad demanda evento_infra evento_zona
1       1     2019-01-01  935      1039    1            1          
2       2     2019-01-01 1067       952    0            1          
3       3     2019-01-01 1030      1147    1            1          
4       4     2019-01-01 1010       900    0            0          
5       5     2019-01-01 1016      1125    0            1          
6       6     2019-01-01  971       865    0            1          
7       7     2019-01-01  994       886    0            0          
8       8     2019-01-01 1023      1136    1            1          
9       9     2019-01-01 1010      1121    1            0          
10     10     2019-01-01 1001       889    0            0          
11     11     2019-01-01 1001      1110    1            0          
12     12     2019-01-01  990       884    0            1          
13     13     2019-01-01  988      1100    1            1          
14     14     2019-01-01 1037       925    0            0          
15     15     2019-01-01  997       889    0            1          
16     16     2019-01-01  967       859    1            0          
17     17     2019-01-01  978       867    0            0          
18     18     2019-01-01 1034      1144    0            1          
19     19     2019-01-01 1046      1157    0            1          
20     20     2019-01-01 1058       943    0            1          
21     21     2019-01-01  963       853    0            1          
22     22     2019-01-01 1006      1115    1            1          
23     23     2019-01-01  983      1090    1            1          
24     24     2019-01-01 1079       962    0            1          
25     25     2019-01-01 1025      1138    1            0          
26     26     2019-01-01 1020       908    0            0          
27     27     2019-01-01 1025      1139    1            0          
28     28     2019-01-01 1014       902    0            1          
29     29     2019-01-01 1012       902    0            0          
30     30     2019-01-01  985       876    0            0          
...    ...    ...        ...       ...     ...          ...        
415341 1109   2019-12-31  995      1101    1            0          
415342 1110   2019-12-31 1033      1149    1            1          
415343 1111   2019-12-31  995      1105    1            0          
415344 1112   2019-12-31  985       875    1            0          
415345 1113   2019-12-31  998       890    0            0          
415346 1114   2019-12-31 1016       904    0            0          
415347 1115   2019-12-31 1025       914    0            1          
415348 1116   2019-12-31 1024       912    1            0          
415349 1117   2019-12-31 1000      1107    1            0          
415350 1118   2019-12-31  952       847    0            0          
415351 1119   2019-12-31  917       813    0            0          
415352 1120   2019-12-31 1027       912    0            1          
415353 1121   2019-12-31 1010       899    1            0          
415354 1122   2019-12-31 1016       904    1            1          
415355 1123   2019-12-31  971       863    0            1          
415356 1124   2019-12-31 1003       893    0            1          
415357 1125   2019-12-31  963      1072    1            1          
415358 1126   2019-12-31  996      1107    1            0          
415359 1127   2019-12-31  900      1002    1            1          
415360 1128   2019-12-31 1027       911    0            1          
415361 1129   2019-12-31 1021      1129    1            0          
415362 1130   2019-12-31 1028       914    0            1          
415363 1131   2019-12-31  990       880    0            0          
415364 1132   2019-12-31  984       874    1            0          
415365 1133   2019-12-31 1022      1133    1            1          
415366 1134   2019-12-31  945      1051    1            0          
415367 1135   2019-12-31  977       871    0            0      

## Open Data

No aplica

## <font color='green'>Data Save</font>

Este proceso, puede copiarse y repetirse en aquellas partes del notebbok que necesiten guardar datos.
Recuerde cambiar las cadenas añadida del fichero para diferenciarlas

1. Infraestructuras con su identificador único

<font color='tomato'> Identificamos los datos a guardar</font>

In [14]:
data_to_save_01 <- tdata_01


<font color='tomato'>Estructura de nombre de archivos:</font>

* Código del caso de uso, por ejemplo "CU_04"
* Número del proceso que lo genera, por ejemplo "_05".
* Número de la tarea que lo genera, por ejemplo "_01"
* En caso de generarse varios ficheros en la misma tarea, llevarán _01 _02 ... después
* Nombre: identificativo de "properData", por ejemplo "_zonasgeo"
* Extensión del archivo

Ejemplo: "CU_04_05_01_01_zonasgeo.json, primer fichero que se genera en la tarea 01 del proceso 05 (Data Collection) para el caso de uso 04 (vacunas)

Importante mantener los guiones bajos antes de proceso, tarea, archivo y nombre

### Proceso 05

In [15]:
caso <- "CU_18"
proceso <- '_05'
tarea <- "_19"
archivo <- "_01"
proper <- "_infraestructuras"
extension <- ".csv"

<font color='tomato'><b> OPCION A:</b> Uso del paquete "tcltk" para mayor comodidad</font>

* Buscar carpeta, escribir nombre de archivo SIN extensión (se especifica en el código)
* Especificar sufijo2 si es necesario
* Cambiar datos por datos_xx si es necesario

In [16]:
# file_save_01 <- paste0(caso, proceso, tarea, tcltk::tkgetSaveFile(), proper, extension) 
# path_out_01 <- paste0(oPath, file_save_01)
# write_csv(data_to_save_01, path_out_01)

# cat('File saved as: ')
# path_out

<font color='tomato'><b> OPCION B:</b> Especificar el nombre de archivo</font>

- Los ficheros de salida del proceso van siempre a Data/Output/.  

In [17]:
file_save_01 <- paste0(caso, proceso, tarea, archivo, proper, extension) 
path_out_01 <- paste0(oPath, file_save_01)
write_csv(data_to_save_01, path_out_01)

cat('File saved as: ')
path_out_01

File saved as: 

[1] "Data/Output/CU_18_05_19_01_infraestructuras.csv"

#### Copia del fichero a Input

Si el archivo se va a usar en otros notebooks, copiar a la carpeta Input

In [18]:
# path_in <- paste0(iPath, file_save)
# file.copy(path_out, path_in, overwrite = TRUE)

2. Diario de infraestructuras

<font color='tomato'> Identificamos los datos a guardar</font>

In [19]:
data_to_save_02 <- tdata_02


<font color='tomato'>Estructura de nombre de archivos:</font>

* Código del caso de uso, por ejemplo "CU_04"
* Número del proceso que lo genera, por ejemplo "_05".
* Número de la tarea que lo genera, por ejemplo "_01"
* En caso de generarse varios ficheros en la misma tarea, llevarán _01 _02 ... después
* Nombre: identificativo de "properData", por ejemplo "_zonasgeo"
* Extensión del archivo

Ejemplo: "CU_04_05_01_01_zonasgeo.json, primer fichero que se genera en la tarea 01 del proceso 05 (Data Collection) para el caso de uso 04 (vacunas)

Importante mantener los guiones bajos antes de proceso, tarea, archivo y nombre

### Proceso 05

In [20]:
archivo <- "_02"
proper <- "_diario_infra"
extension <- ".csv"

<font color='tomato'><b> OPCION A:</b> Uso del paquete "tcltk" para mayor comodidad</font>

* Buscar carpeta, escribir nombre de archivo SIN extensión (se especifica en el código)
* Especificar sufijo2 si es necesario
* Cambiar datos por datos_xx si es necesario

In [21]:
# file_save_02 <- paste0(caso, proceso, tarea, tcltk::tkgetSaveFile(), proper, extension) 
# path_out_02 <- paste0(oPath, file_save_02)
# write_csv(data_to_save_01, path_out_02)

# cat('File saved as: ')
# path_out

<font color='tomato'><b> OPCION B:</b> Especificar el nombre de archivo</font>

- Los ficheros de salida del proceso van siempre a Data/Output/.  

In [22]:
file_save_02 <- paste0(caso, proceso, tarea, archivo, proper, extension) 
path_out_02 <- paste0(oPath, file_save_02)
write_csv(data_to_save_02, path_out_02)

cat('File saved as: ')
path_out_02

File saved as: 

[1] "Data/Output/CU_18_05_19_02_diario_infra.csv"

#### Copia del fichero a Input

Si el archivo se va a usar en otros notebooks, copiar a la carpeta Input

In [23]:
path_in_02 <- paste0(iPath, file_save_02)
file.copy(path_out_02, path_in_02, overwrite = TRUE)

[1] TRUE

## <font color=' #2874a6 '>Main Conclusions</font>

<font color='steelblue'>List and describe the general conclusions of the analysis carried out​.</font>

### Prerequisites

Para que funcione este código se necesita:

* Las rutas de archivos `Data/Input` y `Data/Output` deben existir (relativas a la ruta del _notebook_)
* El paquete tcltk instalado para seleccionar archivos interactivamente. No se necesita en producción.
* Los paquetes tcltk, readr, dplyr deben estar instalados.


### Configuration Management

This notebook has been tested with the following versions of R and packages. It cannot be assured that later versions work in the same way:

* R 4.2.2
* tcltk 4.2.2
* readr 2.1.3
* dplyr 1.0.10


### Data structures

#### Objeto `tdata_01`

- Los datos de origen son los de infraestructuras pero sin indizar
- Los datos de salida son los mismos pero con el índice de fila para unir al diario


#### Objeto `tdata_02`

- Los datos de salida son 415370 filas de las columnas:
    * id_inf
    * fecha
    * capacidad
    * demanda
    * evento_infra
    * evento_zona


#### Observaciones generales sobre los datos

* Los datos han sido generados de forma aleatoria y arbitraria

### Consideraciones para despliegue en piloto

* Es posible que los modelos no arrojen resultados esclarecedores al ser datos _fake_

### Consideraciones para despliegue en producción

* Se deben crear los procesos ETL en producción necesarios para que los datos de entrada estén actualizados
* Se deberían tener datos reales para la puesta en producción

## <font color=' #2874a6 '>Main Actions</font>

#### <font color=' #2874a6 '>Acciones done</font>

<font color='steelblue'>Indicate the actions that have been carried out in this process</font>

- Se han simulado datos de ocupación, capacidad y vivienda

#### <font color=' #2874a6 '>Acctions to perform</font>

<font color='steelblue'>Indicate the actions that must be carried out in subsequent processes</font>

- Se debe añadir la información meteorológica de cada día en cada infraestructura

## <font color='RED'>CODE TO DEPLOY (PILOT)</font>

A continuación se incluirá el código que deba ser llevado a despliegue para producción, dado que se entiende efectúa operaciones necesarias sobre los datos en la ejecución del prototipo

<font color=' red '><b>Description</b></font>

- No hay nada que desplegar en el piloto, ya que estos datos son estáticos o en todo caso cambian con muy poca frecuencia, altamente improbable durante el proyecto.


<font color=' red '><b>CODE</b></font>

In [24]:
# incluir código